provider/pkg/provider/provider.go

```diff
@@ -220,6 +248,60 @@ func (k *commandProvider) Diff(ctx context.Context, req *pulumirpc.DiffRequest)
 	}, nil
 }
 
+func decodeSecret(s *resource.Secret) interface{} {
+	if s == nil {
+		return nil
+	}
+	e := s.Element
+	return e.V
+}
+
+func decodeAny(inputs interface{}) interface{} {
+	switch valT := inputs.(type) {
+	case map[string]interface{}:
+		return decodeMap(valT)
+	case []interface{}:
+		return decodeList(valT)
+	case *resource.Secret:
+		return decodeSecret(valT)
+	}
+	return inputs
+}
+
+func decodeList(inputs []interface{}) []interface{} {
+	decoded := make([]interface{}, 0, len(inputs))
+	for _, val := range inputs {
+		decoded = append(decoded, decodeAny(val))
+	}
+	return decoded
+}
+
+func decodeMap(inputs map[string]interface{}) map[string]interface{} {
+	decoded := make(map[string]interface{})
+	for key, val := range inputs {
+		decoded[key] = decodeAny(val)
+	}
+	return decoded
+}
+
+func decodeConnection(decoded map[string]interface{}) error {
```

Comment:
I'm not sure why connection is special. It would be great if there could be an explanation in the comments.